### PR TITLE
Add hyperparameters into decision forest classification inference algorithm

### PR DESCRIPTION
**Pull request by Vika-F**
This PR continues the work started in the PR https://github.com/oneapi-src/oneDAL/pull/2638 

Co-authored-by: Andreas Huber @ahuber21

### CODE REVIEW

**Comment by Vika-F** [comment]
Green pre-commit: http://intel-ci.intel.com/eee6a7f2-c111-f188-91d9-a4bf010d0e2e

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 1]
```diff
@@ -242,29 +265,66 @@ PredictKernel<algorithmFPType, method, cpu>::~PredictKernel()
...
+    if (hyperparameter != nullptr)
+    {
+        DAAL_INT64 blockSizeValue = 0l;
+        Status st                 = hyperparameter->find(blockSize, blockSizeValue);
+        DAAL_CHECK(0l < blockSizeValue, services::ErrorHyperparameterBadValue);
+        if (cpu == avx512)
+        {
+            // TODO: Move this check into decision_forest::detail::infer_ops::check_parameters_ranges.
+            //       Now the information about enabled CPU instructions is not available in infer_ops.
+            DAAL_CHECK(blockSizeValue % (512 / (8 * sizeof(algorithmFPType))) == 0, services::ErrorHyperparameterBadValue);
```
Comment (razdoburdin) [inline cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i:287]: Is it possible to give some meaningful names to magic constants?

File: cpp/oneapi/dal/algo/decision_forest/infer_types.hpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
-* Copyright 2020 Intel Corporation
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/decision_forest/infer_types.hpp:2]: is it necessary change, due to this file was created before?

File: cpp/oneapi/dal/algo/decision_forest/infer_types.cpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
-* Copyright 2020 Intel Corporation
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/decision_forest/infer_types.cpp:2]: the same comment here

File: cpp/oneapi/dal/algo/decision_forest/detail/infer_ops.hpp [hunk 4]
```diff
@@ -50,14 +58,41 @@ struct infer_ops {
                               const input_t& input,
                               const result_t& result) const {}
 
+    /// Check that the hyperparameters of the algorithm belong to the expected ranges
+    void check_parameters_ranges(const param_t& params, const input_t& input) const {
+        ONEDAL_ASSERT(params.get_block_size() > 0x0l);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/decision_forest/detail/infer_ops.hpp:63]: should this checks be not in ONEDAL_ASSERT, due to it works only with dbg flag? Its common question for all hp algorithms. Probably we can doublecheck in examples/tests

File: cpp/oneapi/dal/algo/decision_forest/detail/infer_ops.cpp [hunk 5]
```diff
@@ -15,20 +15,45 @@
...
+
+    infer_result<Task> operator()(const Context& context,
+                                  const descriptor_base<Task>& desc,
+                                  const infer_input<Task>& input) const {
+        const auto params = select_parameters(context, desc, input);
+        return implementation(context, desc, params, input);
+    }
+
+private:
+    inline auto implementation(const Context& context,
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/decision_forest/detail/infer_ops.cpp:50]: Should naming be aligned with other hp algos? for example in cov/linreg it names const Policy& ctx, its little bit confusing

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 6]
```diff
@@ -493,25 +567,27 @@ DAAL_FORCEINLINE void PredictClassificationTask<float, avx512>::predictByTree(co
...
-                                                                               const double * const split_point, double * const resPtr,
-                                                                               const size_t iTree)
+DAAL_FORCEINLINE Status PredictClassificationTask<double, avx512>::predictByTree(const double * const x, const size_t sizeOfBlock, const size_t nCols,
+                                                                                 const featureIndexType * const feat_idx,
+                                                                                 const leftOrClassType * const left_son,
+                                                                                 const double * const split_point, double * const resPtr,
+                                                                                 const size_t iTree)
 {
-    if (sizeOfBlock == _DEFAULT_BLOCK_SIZE)
+    if (sizeOfBlock == _blockSize)
```
Comment (Alexandr-Solovev) [inline cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i:586]: is it necessary to add some checks for _blocksSize somewhere in DAAL part?

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i [hunk 7]
```diff
@@ -128,7 +126,11 @@ public:
           _cachedData(nullptr),
           _cachedModel(nullptr),
           _averageTreeSize(0),
-          _cachedNClasses(0)
+          _cachedNClasses(0),
+          _blockSize(0),
```
Comment (Alexandr-Solovev) [inline cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i:130]: Should the default values be the same as it was previously defined?
#define _DEFAULT_BLOCK_SIZE                      32
#define _DEFAULT_BLOCK_SIZE_COMMON               22
#define _MIN_TREES_FOR_THREADING                 100

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i (hunk 1 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i:287]: Done. Is it better now?

File: cpp/oneapi/dal/algo/decision_forest/infer_types.hpp (hunk 2 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/decision_forest/infer_types.hpp:2]: Thanks for mentioning. Fixed this.

File: cpp/oneapi/dal/algo/decision_forest/detail/infer_ops.hpp (hunk 4 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/decision_forest/detail/infer_ops.hpp:64]: Good point. Agree that those need to be modified to regular checks, not asserts.
But this change will have to be done in LinReg and Covariance also. I do not want to have all those changes part of this PR, rather will do it as a follow up work.

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i (hunk 6 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i:589]: The checks are implemented here:
https://github.com/oneapi-src/oneDAL/pull/2692/files/596158e2645e872ad4232142734dad0eef582c1c#diff-f7c3689a30067e4f500431476e1897d6949edfa783c36c86a86805d4267d6cd3R282

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i (hunk 7 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_predict_dense_default_batch_impl.i]: I've modified; but those values are anyway overridden by a 'setHyperparameters()' call here:
https://github.com/oneapi-src/oneDAL/pull/2692/files/#diff-f7c3689a30067e4f500431476e1897d6949edfa783c36c86a86805d4267d6cd3R314
So, everything works Ok even if no hyperparameters are provided from oneDAL side at all.

**Comment by Vika-F** [comment]
@mergify rebase

**Review by Alexandr-Solovev — APPROVED** [review]

**Pull request merged**